stan/math/opencl/kernel_generator/reduction_2d.hpp

```diff
@@ -0,0 +1,342 @@
+        generated, generated_all, ng, row_index_name, col_index_name);
+
+    parts.args += out_parts.args;
+    parts.reduction_2d += "if (lid_i == 0) {\n"
+                     + result.var_name_
+                     + "_global[wg_id_j * n_groups_i + wg_id_i] = "
+                     + derived().var_name_ + "_local[0];\n"
+                     "}\n";
+    return parts;
+  }
+
+  /**
+   * Generates kernel code for this and nested expressions.
+   * @param row_index_name row index variable name
+   * @param col_index_name column index variable name
+   * @param view_handled whether whether caller already handled matrix view
+   * @param var_name_arg name of the variable in kernel that holds argument to
+   * this expression
+   * @return part of kernel with code for this and nested expressions
+   */
+  inline kernel_parts generate(const std::string& row_index_name,
+                               const std::string& col_index_name,
+                               const bool view_handled,
+                               const std::string& var_name_arg) const {
+    kernel_parts res;
+    res.declarations = "__local " + type_str<Scalar>() + " " + var_name_
+                       + "_local[LOCAL_SIZE_];\n" + type_str<Scalar>() + " "
+                       + var_name_ + " = " + init_ + ";\n";
+    res.body = var_name_ + " = " + Operation::generate(var_name_, var_name_arg)
+               + ";\n";
+    res.reduction_2d =
+          var_name_ + "_local[lid_i] = " + var_name_ + ";\n"
+          "barrier(CLK_LOCAL_MEM_FENCE);\n"
+          "for (int step = lsize_i / REDUCTION_STEP_SIZE; "
+                "step > 0; step /= REDUCTION_STEP_SIZE) {\n"
+          "  if (lid_i < step) {\n"
+          "    for (int i = 1; i < REDUCTION_STEP_SIZE; i++) {\n"
+          "      " + var_name_ + "_local[lid_i] = " +
+        Operation::generate(var_name_ + "_local[lid_i]",
+                            var_name_ + "_local[lid_i + step * i]") + ";\n"
+          "    }\n"
+          "  }\n"
+          "  barrier(CLK_LOCAL_MEM_FENCE);\n"
+          "}\n";
+    return res;
+  }
+
+  /**
+   * Number of rows of a matrix that would be the result of evaluating this
+   * expression.
+   * @return number of rows
+   */
+  inline int rows() const {
+    int arg_rows = this->template get_arg<0>().rows();
+    int arg_cols = this->template get_arg<0>().cols();
+    if (arg_cols == 0) {
+      return 1;
+    }
+    if (arg_cols == -1) {
+      return -1;
+    }
+    return internal::colwise_reduction_wgs_rows(arg_rows, arg_cols);
+  }
+
+  /**
+   * Number of columns of a matrix that would be the result of evaluating this
+   * expression.
+   * @return number of columns
+   */
+  inline int cols() const {
+    int arg_rows = this->template get_arg<0>().rows();
+    int arg_cols = this->template get_arg<0>().cols();
+    if (arg_cols == 0) {
+      return 0;
+    }
+    if (arg_cols == -1) {
+      return -1;
+    }
```

Comment:
Does -1 here just mean all of the columns?

Reply:
kernel generator is using -1 for a not-yet-determined-seize (similar to Eigen::Dynamic). I actually have constant I could use instead of a magic number. Fixed.